python/pyspark/sql/avro/functions.py

```diff
@@ -0,0 +1,134 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+A collections of builtin avro functions
+"""
+
+
+from pyspark import since, SparkContext
+from pyspark.sql.column import Column, _to_java_column
+from pyspark.util import _print_missing_jar
+
+
+@since(3.0)
+def from_avro(data, jsonFormatSchema, options={}):
+    """
+    Converts a binary column of avro format into its corresponding catalyst value. The specified
+    schema must match the read data, otherwise the behavior is undefined: it may fail or return
+    arbitrary result.
+
+    Avro is built-in but external data source module since Spark 2.4. Please deploy the application
+    as per the deployment section of "Apache Avro Data Source Guide".
+
+    :param data: the binary column.
+    :param jsonFormatSchema: the avro schema in JSON string format.
+    :param options: options to control how the Avro record is parsed.
+
+    >>> from pyspark.sql import Row
+    >>> from pyspark.sql.avro.functions import from_avro, to_avro
+    >>> data = [(1, Row(name='Alice', age=2))]
+    >>> df = spark.createDataFrame(data, ("key", "value"))
+    >>> avroDf = df.select(to_avro(df.value).alias("avro"))
+    >>> avroDf.collect()
+    [Row(avro=bytearray(b'\\x00\\x00\\x04\\x00\\nAlice'))]
+    >>> jsonFormatSchema = '''{"type":"record","name":"topLevelRecord","fields":
+    ...     [{"name":"avro","type":[{"type":"record","name":"value","namespace":"topLevelRecord",
+    ...     "fields":[{"name":"age","type":["long","null"]},
+    ...     {"name":"name","type":["string","null"]}]},"null"]}]}'''
+    >>> avroDf.select(from_avro(avroDf.avro, jsonFormatSchema).alias("value")).collect()
+    [Row(value=Row(avro=Row(age=2, name=u'Alice')))]
+    """
+
+    sc = SparkContext._active_spark_context
+    try:
+        jc = sc._jvm.org.apache.spark.sql.avro.functions.from_avro(
+            _to_java_column(data), jsonFormatSchema, options)
+    except TypeError as e:
```

Comment:
@gaborgsomogyi, I think you can check by, for instance, `Utils.classForName("any exposed API of avro class")`, catch Py4j exception, and shows a proper error message.

Reply:
You mean write a scala wrapper to call `from_avro` function? Does it mean you think the existing way in kinesis is wrong somehow?

Reply:
Exiting way is okay - it works. But I was thinking catching `TypeError` with checking string `'JavaPackage' object is not callable` is a bit flaky. So, maybe we could check .. 

```python
from py4j.protocol import Py4JJavaError
...
is_avro_found = spark._jvm.org.apache.spark.util.Utils.classIsLoadable(
    "org.apache.spark.sql.avro.functions")
if not is_avro_found:
    raise RuntimeError(formatted_missing_jar_msg("Avro", "avro", "avro", sc.version))
```

Reply:
You can leave Kinesis's codes as are since this PR doesn't target to fix Kinesis's.  `_print_missing_jar` can be renamed to `_formatted_missing_jar_msg` that returns a string.

Reply:
Hm, but okay. Let's leave it as is and fix it later together since I found other places that use similar approaches.

Reply:
I was looking for this suggestion but seems like we can solve this later. Happy to catch this up in a later jira... Thanks.